### ISSUE TITLE
cls_loss = 0.693147

### ISSUE THREAD

**Issue by zhonhel** [CongWeilin/mtcnn-caffe#42]
When training ,and when the batch is for training roi loss or pts loss, the cls_loss equals a constant which is 0.693147.
I wander if cls_loss != 0 can influence the training?

**Comment by wolf345** [comment]
@zhonhel the same here!

**Comment by BorisLestsov** [comment]
I had the same problem. Look at the weights initialisation at some layers in prototxt. Its Gaussian with very low mean. So when the training is started, weights are almost zero, so the gradients are almost zero too. So the network is not being trained. If you want to solve the problem, go through all the prototxt s and replace those gaussains with "xavier" or smth like this. Worked for me

BTW, scripts in theese repo contain tons of errors.

**Comment by Zouhj** [comment]
@BorisLestsov I have changed it like you described. but the cls_loss = 0.693147 still exists. I don't know whether it will influence the training too. and the result of trained model is not good, especially two regression task. so what the performance of your experiment? can you give me some another advice? Thank you~

**Comment by jianbaba** [comment]
@Zouhj 0.693147不会影响你的训练，0.693147代表当时那个minibatch不是用的分类数据，即没有分类数据传播到损失层。即不训练分类任务时的默认值为0.693147。而且不会反向传递梯度

**Comment by abc20002929** [comment]
@zhonhel when training roi/pts, the input values('conv4-1-valid') of SoftmaxWithLoss Layer(‘cls_loss’) are set to 0 in pythonLayer.py. 
softmax : [exp(0)/((exp(0)+exp(0)), exp(0)/((exp(0)+exp(0))] = [0.5,0.5]
loss : -log(0.5)*2*batch_size/(2*batch_size) = -log(0.5) = 0.693147

**Comment by zhonhel** [comment]
@abc20002929 
Thank you. I later found this principle. So in cls_Layer_fc python layer, we can manually set non-cls samples' label and score. Doing this can make these samples' loss approximatly equal 0.

for examples,
In 12net/pythonLayer.py cls_Layer_fc python layer's forward function, you can set non-cls samples' label and score like this:
`
	top[1].data[non_cls_sample_idx][0]=0                #set its label to #0 class
        top[0].data[non_cls_sample_idx][0][0][0]=10.0   #let it has a big score corresponding to #0 class
        top[0].data[non_cls_sample_idx][1][0][0]=-10.0 #let it has a small score corresponding to #1 class
`
After doing this, you will make non-cls sample's cls loss approximatly equal 0.

**Comment by yungs2017** [comment]
The cls loss is always 0.693147 or 87.3365, and the output loss is nan
```
I0912 17:55:33.864660 11964 sgd_solver.cpp:105] Iteration 20000, lr = 0.001
I0912 17:55:34.774521 11964 solver.cpp:218] Iteration 20500 (549.538 iter/s, 0.909856s/500 iters), loss = nan
I0912 17:55:34.774559 11964 solver.cpp:237]     Train net output #0: cls_Acc = 0.234375
I0912 17:55:34.774567 11964 solver.cpp:237]     Train net output #1: cls_loss = 87.3365 (* 1 = 87.3365 loss)
I0912 17:55:34.774572 11964 solver.cpp:237]     Train net output #2: pts_loss = 0 (* 0.5 = 0 loss)
I0912 17:55:34.774576 11964 solver.cpp:237]     Train net output #3: roi_loss = 0 (* 0.5 = 0 loss)
I0912 17:55:34.774581 11964 sgd_solver.cpp:105] Iteration 20500, lr = 0.001
I0912 17:55:35.697571 11964 solver.cpp:218] Iteration 21000 (541.721 iter/s, 0.922984s/500 iters), loss = nan
I0912 17:55:35.697609 11964 solver.cpp:237]     Train net output #0: cls_Acc = 0
I0912 17:55:35.697618 11964 solver.cpp:237]     Train net output #1: cls_loss = 0.693147 (* 1 = 0.693147 loss)
I0912 17:55:35.697624 11964 solver.cpp:237]     Train net output #2: pts_loss = 0 (* 0.5 = 0 loss)
I0912 17:55:35.697654 11964 solver.cpp:237]     Train net output #3: roi_loss = nan (* 0.5 = nan loss)
I0912 17:55:35.697657 11964 sgd_solver.cpp:105] Iteration 21000, lr = 0.001
I0912 17:55:36.599164 11964 solver.cpp:218] Iteration 21500 (554.614 iter/s, 0.901528s/500 iters), loss = nan
I0912 17:55:36.599205 11964 solver.cpp:237]     Train net output #0: cls_Acc = 0.265625
I0912 17:55:36.599213 11964 solver.cpp:237]     Train net output #1: cls_loss = 87.3365 (* 1 = 87.3365 loss)
I0912 17:55:36.599218 11964 solver.cpp:237]     Train net output #2: pts_loss = 0 (* 0.5 = 0 loss)
I0912 17:55:36.599223 11964 solver.cpp:237]     Train net output #3: roi_loss = 0 (* 0.5 = 0 loss)
```
no one helps?
